src/eckit/stats/detail/Counter.cc

```diff
@@ -0,0 +1,166 @@
+/*
+ * (C) Copyright 1996- ECMWF.
+ *
+ * This software is licensed under the terms of the Apache Licence Version 2.0
+ * which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * In applying this licence, ECMWF does not waive the privileges and immunities
+ * granted to it by virtue of its status as an intergovernmental organisation nor
+ * does it submit to any jurisdiction.
+ */
+
+
+#include "eckit/stats/detail/Counter.h"
+
+#include <limits>
+#include <ostream>
+
+#include "eckit/config/Parametrisation.h"
+#include "eckit/exception/Exceptions.h"
+#include "eckit/stats/Field.h"
+
+
+namespace eckit::stats::detail {
+
+
+static double get_param(const Parametrisation& param, const std::string& key, double def) {
+    double value(def);
+    param.get(key, value);
+    return value;
+}
+
+
+Counter::Counter(double missingValue, bool hasMissing, double lowerLimit, double upperLimit) :
+    count_(0),
+    missing_(0),
+    countBelowLowerLimit_(0),
+    countAboveUpperLimit_(0),
+    minIndex_(0),
+    maxIndex_(0),
+    missingValue_(missingValue),
+    lowerLimit_(lowerLimit),
+    upperLimit_(upperLimit),
+    min_(std::numeric_limits<double>::quiet_NaN()),
+    max_(std::numeric_limits<double>::quiet_NaN()),
+    hasMissing_(hasMissing),
+    hasLowerLimit_(lowerLimit_ == lowerLimit_),
```

Comment:
This looks odd. I assume it is essentially testing if things are NaN, or similar. But I worry that the equality comparison could be optimised away to true.

Reply:
At this level, to support the statistics, they operate on real values and NaN if not assigned/not-definable (there's no "finite missing value"). The assumption here is the NaN support -- and you're right, this can be compiled out. I know that MIR and Atlas will fail in specific places to to similar optimisations. Maybe I should protect this with a static_assert? But the problem is widespread.